### PR TITLE
Augment the Abuse Report API with new fields

### LINKED ISSUE #10643: Augment abuse API with new fields for add-on reporting within Firefox
PRD: https://docs.google.com/document/d/1K5Th9UmpayfVMXadkxU6cDO3sFKOwYozQgyoN0WEGwY/edit?ts=5c4b6e8d#heading=h.af3x6jga7lsy

The Firefox Add-ons Manager is going to implement an abuse report feature similar to the existing one on AMO. To that end we need to add some fields to the [current abuse API](https://addons-server.readthedocs.io/en/latest/topics/api/abuse.html) to store additional data that can be sent by Firefox. The PRD specifies we want the following fields (need to sort out how to reconcile this with the current API):

- Client ID hash
- Add-on GUID
- Add-on name
- Add-on summary (description field in manifest)
- Add-on version
- Signature state (curated, non-curated, partner, unsigned, temporary installation)
- Application (Firefox, Firefox for Android)
- Application version
- Application locale
- Operating system
- Operating system version
- Install date
- Abuse report flags (malware, spam, broken search, broken new tab, broken websites, etc.)
- Abuse report text
- Region (tbd - fwiw the current model stores the ip address directly)
- Install origin
- Install method
- Entry point (post-uninstall prompt, options menu, context menu, etc.)

### PR DESCRIPTION
**Pull request by diox**
Those new fields are meant to be used for add-on reporting from within Firefox itself.
Includes a refactor of the abuse API internals to make adding those fields a lot easier.

Fixes #10643

### CODE REVIEW

File: docs/topics/api/abuse.rst [hunk 1]
```diff
@@ -19,14 +19,28 @@ Submitting an add-on abuse report
...
+    :<json string|null addon_signature: The add-on signature state. Accepted values: TBD.
+    :<json string|null addon_summary: The add-on summary in the locale used by the client.
+    :<json string|null addon_version: The add-on version string.
+    :<json string|null application: The application used by the client. Can be either ``firefox`` or ``android``.
+    :<json string|null application_locale: The locale used by the client for the application.
+    :<json string|null client_id: The client's hashed telemetry ID.
+    :<json string|null install_date: The add-on install date.
+    :<json string|null operating_system: The client's operating system.
+    :<json string|null operating_system_version: The client's operating system version.
+    :<json string|null reason: The reason for the report. Can be ``malware``, ``spam_or_advertising``, ``search_takeover``, ``new_tab_takeover``, `breaks_website``, `offensive``, `does_not_match_description`` or ``does_not_work``.
```
Comment (EnTeQuAk) [inline docs/topics/api/abuse.rst]: at ``breaks_website`` you missed a "`", same for ``offensive`` and `does_not_match_description``

File: docs/topics/api/abuse.rst [hunk 2]
```diff
@@ -37,6 +51,20 @@ to if nessecary.
     :>json int|null addon.id: The add-on id on AMO. If the guid submitted didn't match a known add-on on AMO, then null.
     :>json string|null addon.slug: The add-on slug. If the guid submitted didn't match a known add-on on AMO, then null.
     :>json string message: The body/content of the abuse report.
+    :>json string|null addon_install_entry_point: The add-on install entry point.
```
Comment (EnTeQuAk) [inline docs/topics/api/abuse.rst]: These (and below) look duplicated?

File: src/olympia/abuse/models.py [hunk 3]
```diff
@@ -14,6 +14,41 @@
...
+        None: None,
+        1: 'malware',
+        2: 'spam_or_advertising',
+        3: 'search_takeover',
+        4: 'new_tab_takeover',
+        5: 'breaks_websites',
+        6: 'offensive',
+        7: 'does_not_match_description',
+        8: 'does_not_work',
+    }
```
Comment (EnTeQuAk) [inline src/olympia/abuse/models.py]: I wonder if we should be using constants for values `None` to `8` but as this may change back and forth with the PRD I think it's fine for now. Maybe keep that in mind?

File: src/olympia/abuse/models.py [hunk 4]
```diff
@@ -33,14 +68,51 @@ class AbuseReport(ModelBase):
         on_delete=models.SET_NULL)
     message = models.TextField()
 
+    # Extra optional fields for more information, giving some context that is
+    # meant to be extracted automatically by the client (i.e. Firefox) and
+    # submitted via the API.
+    client_id = models.CharField(
+        default=None, max_length=64, blank=True, null=True)
+    addon_name = models.CharField(
+        default=None, max_length=255, blank=True, null=True)
```
Comment (EnTeQuAk) [inline src/olympia/abuse/models.py]: Are these so long max lengths because we usually don't care? Addon name is currently limited to 50 characters in the devhub form for example.

File: src/olympia/abuse/models.py [hunk 5]
```diff
@@ -33,14 +68,51 @@ class AbuseReport(ModelBase):
...
+        null=True)
+
     class Meta:
         db_table = 'abuse_reports'
 
     def send(self):
         if self.reporter:
             user_name = '%s (%s)' % (self.reporter.name, self.reporter.email)
         else:
-            user_name = 'An anonymous coward'
```
Comment (EnTeQuAk) [inline src/olympia/abuse/models.py]: wtf :fearful: (don't look at git blame, it hurts)

File: src/olympia/abuse/models.py [hunk 6]
```diff
@@ -50,6 +122,12 @@ def send(self):
         send_mail(
             six.text_type(self), msg, recipient_list=(settings.ABUSE_EMAIL,))
 
+    def save(self, *args, **kwargs):
+        creation = not self.pk
+        super(AbuseReport, self).save(*args, **kwargs)
+        if creation:
+            self.send()
```
Comment (EnTeQuAk) [inline src/olympia/abuse/models.py]: I always wondered if a `post_save` might be cleaner as we already get the information if we are creating or updating and Django can determine that on it's own :thinking:

File: src/olympia/abuse/serializers.py [hunk 7]
```diff
@@ -1,16 +1,79 @@
...
+        if target_name not in data:
+            msg = serializers.Field.default_error_messages['required']
+            raise serializers.ValidationError({
+                target_name: [msg]
+            })
+
+    def to_internal_value(self, data):
+        output = super(BaseAbuseReportSerializer, self).to_internal_value(data)
+        request = self.context['request']
+        output['ip_address'] = request.META.get('REMOTE_ADDR')
```
Comment (EnTeQuAk) [inline src/olympia/abuse/serializers.py]: We should have `ip_address` defined in the serializer so that it gets validated and we don't store unvalidated input.

File: src/olympia/abuse/tests/test_views.py [hunk 8]
```diff
@@ -138,6 +140,102 @@ def test_throttle(self):
...
+        for field in ('message', 'client_id', 'addon_name', 'addon_summary',
+                      'addon_version', 'application_version',
+                      'application_locale', 'operating_system',
+                      'addon_install_origin'):
+            assert getattr(report, field) == data[field], field
+        # More complex comparisons:
+        assert report.addon_signature is None
+        assert report.application == amo.FIREFOX.id
+        assert report.install_date == datetime(2004, 8, 15, 16, 23, 42)
+        assert report.reason == 2  # Spam / Advertising
```
Comment (EnTeQuAk) [inline src/olympia/abuse/tests/test_views.py]: Really should be constants :laughing:

File: docs/topics/api/abuse.rst (hunk 2 above)
Comment (diox) [reply docs/topics/api/abuse.rst]: Those are for the output, above is for the input

File: src/olympia/abuse/models.py (hunk 3 above)
Comment (diox) [reply src/olympia/abuse/models.py]: Yeah... none of the choices are final at this point so I didn't want to go too far with the constants yet... Also, we really should be using something like https://github.com/twidi/django-extended-choices

File: src/olympia/abuse/models.py (hunk 4 above)
Comment (diox) [reply src/olympia/abuse/models.py]: The add-on doesn't have to be listed, and it could have been installed from before the limits were put in place in devhub anyway. I didn't have any direction on this so I just picked the standard largish value.

File: src/olympia/abuse/models.py (hunk 6 above)
Comment (diox) [reply src/olympia/abuse/models.py]: Yeah, I always struggle with this choice. Signals can be hard to follow, so I tend to try to override `save` these days...

File: src/olympia/abuse/serializers.py (hunk 7 above)
Comment (diox) [reply src/olympia/abuse/serializers.py]: The problem is we don't want it to be submitted directly. And we don't want to error on it either. So I think it's the best compromise. It will go away in https://github.com/mozilla/addons-server/issues/10720 anyway.

File: docs/topics/api/abuse.rst (hunk 2 above)
Comment (EnTeQuAk) [reply docs/topics/api/abuse.rst]: Eh, indeed… `<` vs `>` - sorry, I didn't see that

File: src/olympia/abuse/models.py (hunk 3 above)
Comment (EnTeQuAk) [reply src/olympia/abuse/models.py]: > django-extended-choices

This sounds great, let's do this at some point. Might cleanup our code a lot

File: src/olympia/abuse/models.py (hunk 4 above)
Comment (EnTeQuAk) [reply src/olympia/abuse/models.py]: Sounds fair :+1:

File: src/olympia/abuse/serializers.py (hunk 7 above)
Comment (EnTeQuAk) [reply src/olympia/abuse/serializers.py]: Ah, if it goes away (I forgot) that's fine then.
